pkg/networking/backend_sg_provider.go

```diff
@@ -0,0 +1,245 @@
+package networking
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	ec2sdk "github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	networking "k8s.io/api/networking/v1beta1"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/aws/services"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	resourceTypeSecurityGroup = "security-group"
+	tagKeyK8sCluster          = "elbv2.k8s.aws/cluster"
+	tagKeySGType              = "elbv2.k8s.aws/type"
+	tagValueBackend           = "backend"
+
+	explicitGroupFinalizerPrefix = "group.ingress.k8s.aws/"
+	implicitGroupFinalizer       = "ingress.k8s.aws/resources"
+
+	sgDescription = "[k8s] Shared Backend SecurityGroup for LoadBalancer"
+)
+
+// BackendSGProvider is responsible for providing backend security groups
+type BackendSGProvider interface {
+	// Get returns the backend security groups to use
+	Get(ctx context.Context) ([]string, error)
+	// Release cleans up the auto-generated backend SG if necessary
+	Release(ctx context.Context) error
+}
+
+// NewBackendSGProvider constructs a new  defaultBackendSGProvider
+func NewBackendSGProvider(clusterName string, backendSGEnabled bool, backendSGs []string, vpcID string,
+	ec2Client services.EC2, k8sClient client.Client, logger logr.Logger) *defaultBackendSGProvider {
+	var sanitizedBackendSGs []string
+	for _, sg := range backendSGs {
```

Comment:
i prefer to validate the format strictly and early in https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/config/controller_config.go#L100
so we don't do the trimp and validate that if any of the specified sgIDs don't match the "sg-xxxxx" regex format, we fail.

Reply:
just the regex match sg-xxx, or actually query the EC2 configuration? I'm thinking about a less strict validation where we accept sg name as well, and validate only when the actual resource gets created. This gives the opportunity to create the sg resources later. Validation would make sense for the sg-xxx format since it is less likely the exact value to be created later.